core/server/common/src/main/java/alluxio/master/StateLockManager.java

```diff
@@ -0,0 +1,247 @@
+
+  /** Whether exclusive locking will trigger interrupt-cycle. */
+  private boolean mInterruptCycleEnabled;
+  /** Interval at which threads around shared-lock will be interrupted during interrupt-cycle. */
+  private long mInterruptCycleInterval;
+  /** Used to synchronize execution/termination of interrupt-cycle. */
+  private Lock mInterruptCycleLock = new ReentrantLock(true);
+  /** How many active exclusive locking attempts. */
+  private volatile long mInterruptCycleRefCount = 0;
+  /** The future for the active interrupt cycle. */
+  private ScheduledFuture<?> mInterrupterFuture;
+
+  /** This is the deadline for forcing the lock. */
+  private long mForcedDurationMs;
+
+  // TODO(ggezer): Make it bound to a process start/stop cycle.
+  /** Shared locking requests will fail until this time. */
+  private long mExclusiveOnlyDeadlineMs = -1;
+
+  /**
+   * Creates a new state-lock manager.
+   */
+  public StateLockManager() {
+    mSharedWaitersAndHolders = new ConcurrentHashSet<>();
+    // Init members.
+    mScheduler = Executors
+        .newSingleThreadScheduledExecutor(ThreadFactoryUtils.build("state-lock-manager-%d", true));
+    // Read properties.
+    mInterruptCycleEnabled = ServerConfiguration
+        .getBoolean(PropertyKey.MASTER_BACKUP_STATE_LOCK_INTERRUPT_CYCLE_ENABLED);
+    mInterruptCycleInterval =
+        ServerConfiguration.getMs(PropertyKey.MASTER_BACKUP_STATE_LOCK_INTERRUPT_CYCLE_INTERVAL);
+    mForcedDurationMs =
+        ServerConfiguration.getMs(PropertyKey.MASTER_BACKUP_STATE_LOCK_FORCED_DURATION);
+    // Validate properties.
+    Preconditions.checkArgument(mInterruptCycleInterval > 0,
+        "Interrupt-cycle interval should be greater than 0.");
+  }
+
+  /**
+   * This is called by owning process in order to signal that
+   * the state is read completely and masters are started.
+   *
+   * This triggers the beginning of exclusive-only maintenance mode for the state-lock.
+   * Note: Calling it multiple times does not reset the maintenance window.
+   */
+  public void mastersStartedCallback() {
+    if (mExclusiveOnlyDeadlineMs == -1) {
+      mExclusiveOnlyDeadlineMs = System.currentTimeMillis()
+          + ServerConfiguration.getMs(PropertyKey.MASTER_BACKUP_STATE_LOCK_EXCLUSIVE_DURATION);
+    }
+  }
+
+  /**
+   * Locks the state shared.
+   *
+   * Calling thread might be interrupted by this manager,
+   * if it found to be waiting for the shared lock under when:
+   *  - backup is exiting grace-cycle and entering the lock permanently
+   *  - backup is in progress
+   *
+   * @return the lock resource
+   * @throws InterruptedException
+   */
+  public LockResource lockShared() throws InterruptedException {
+    // Do not allow taking shared lock during safe-mode.
+    if (System.currentTimeMillis() < mExclusiveOnlyDeadlineMs) {
+      String safeModeMsg = String.format(
+          "Master still in exclusive-only phase for the state-lock."
+              + "Please see documentation for %s.",
```

Comment:
Would be helpful to state how much longer the deadline is

Reply:
Added more traces https://github.com/Alluxio/alluxio/pull/11535/commits/b8724736bbe4463dadc8920db951ab867fb3cd4f